clang/include/clang/Basic/FileManager.h

```diff
@@ -272,18 +272,21 @@ class FileManager : public RefCountedBase<FileManager> {
   /// MemoryBuffer if successful, otherwise returning null.
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
   getBufferForFile(const FileEntry *Entry, bool isVolatile = false,
-                   bool RequiresNullTerminator = true);
+                   bool RequiresNullTerminator = true,
+                   Optional<cas::ObjectRef> *CASContents = nullptr);
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>>
   getBufferForFile(StringRef Filename, bool isVolatile = false,
-                   bool RequiresNullTerminator = true) {
+                   bool RequiresNullTerminator = true,
```

Comment:
I am not sure if I like that you get the ObjectRef when you getBuffer. Maybe use a new API that returns `ErrorOr<ObjectRef>`, or add a new API that retrieve both with a different name if that is what we always doing.

Reply:
The general idea with `getBufferForFile()` returning both is that after you open a file you'd get both the buffer and the `ObjectRef` for the buffer for "free", so you'd avoid 2 lookups if you need both of them.

I can see the usefulness of adding a new API to return an `ObjectRef` only, but do you have a concern with adding the optional parameter to `getBufferForFile()`? If I add an extra set of `getBufferAndObjectRefForFile()` functions it will be essentially the same signatures as the existing `FileManager::getBufferForFile()` functions but with an added parameter, in which case it seems natural to just add it as an optional parameter.

Reply:
Yes, I can understand that, which is why I made two proposals depending on your use case. But now with a second thought, I wonder if this is the correct way to model CASFileSystem.

Maybe CASID or ObjectRef should not be the product after opening the file, but more like `status` that you can get without opening the file? Then you can cache the ObjectRef in the FileSystemStatCache if needed so using another call to get ObjectRef is not expensive.

Reply:
> Maybe CASID or ObjectRef should not be the product after opening the file, but more like status that you can get without opening the file?

I'm a bit concerned that "ingestion to get ObjectRef", whenever it happens for `CASProvidingFileSystem`, would do relatively significant work (open the file, get the buffer and ingest it), while "get status" is generally expected to be way more lightweight than that, so there'd be a disconnect that I'd prefer to avoid.

There's possibility in the future for a `vfs::FileSystem` implementation where getting `ObjectRef` is as cheap as "get status" but we could re-evaluate if the FileSystem/FileManager APIs would benefit from another change if this comes to be.

Reply:
I guess we are on the same page? My point was that we should make `getCASContentsForFile` cheap instead of getting it for `free` when you open it for reading. Currently, it requires hashing so it should just be a field stored in `CASProvidingFile`.

Also `getCASContentsForFile` is not a good name. What you get is a CAS ObjectRef, not the content of the file.

Reply:
> I guess we are on the same page? My point was that we should make getCASContentsForFile cheap instead of getting it for free when you open it for reading. Currently, it requires hashing so it should just be a field stored in CASProvidingFile.

You suggested "more like `status` that you can get without opening the file" and the thing I wanted to convey is that it shouldn't be the case that just by calling `vfs::FileSystem::status()` it should give you the `ObjectRef` for the file buffer, because the act of getting the `ObjectRef` opens the file and reads its buffer, which is way more expensive than what `status()` is expected to do.

Also it's not clear to me what specific source change you are suggesting, could you elaborate? You suggested "should just be a field stored in CASProvidingFile" and "instead of getting it for free when you open it for reading" but these are contradictory because to get the `CASProvidingFile` you need to call `vfs::FileSystem::openFileForRead()` so it is still the case of "getting it for free when you open it for reading" and it won't be "more like `status`".

> Also getCASContentsForFile is not a good name. What you get is a CAS ObjectRef, not the content of the file.

How about `getObjectRefForContentsOfFile`?

Reply:
I am suggesting `CASProvidingFile` to have a field `ObjectRef` so it would be cheap when you try to get ObjectRef once you have the file created. Yes, it currently requires openFile but that is implementation details and it might not be needed if the file system is CAS backed. The point is:
* You should be able to get ObjectRef without openFile (from API's point of view)
* It should not be more expensive to get ObejctRef and openFile in two API calls.

> How about getObjectRefForContentsOfFile?

That seems long. `getObjectRefForContent` is probably good enough.

Reply:
> You should be able to get ObjectRef without openFile (from API's point of view)

So this is the `vfs::FileSystem::getCASContentsForFile()` function in this PR, right?

> It should not be more expensive to get ObejctRef and openFile in two API calls.

I'd prefer to keep the implementation of `CASProvidingFileSystem` as simple as possible for now (instead of have it do its own ad-hoc caching for its queries) since its primary use-case is to be the underlying file-system of `DependencyScanningWorkerFilesystem` which already does its own caching, and any other CAS-related file-systems can implement the `vfs::File::getCASContents()` function with their own data/caches.

Later on, if there is a need, we can add caching on top or introduce something like `CachingCASProvidingFileSystem` or just compose `DependencyScanningWorkerFilesystem`+`CASProvidingFileSystem` for a file-system that both provides `ObjectRef` for contents and also caches  openFile calls.

> That seems long. getObjectRefForContent is probably good enough.

SGTM 👍